### PR TITLE
Refactors language holder

### LINKED ISSUE #25663: pAIs with universal translator taking control of bots, do not keep their languages

### LINKED ISSUE #27087: monkeys speak galcom
VERY fun but oh well, it is a bug.

### LINKED ISSUE #29568: MMI's have UI to change languages, but are locked into a language after being put into a mech.
Funny, but annoying.

### LINKED ISSUE #37690: Aphasia is not reset on cloning like other severe traumas are.
And i'm pretty sure brain surgery won't fix it after being cloned so enjoy speaking gibberish forever.

### LINKED ISSUE #42871: Voice Change symptom permanently prevents speaking
## Round ID:

102365

## Testmerges:
None

## Reproduction:
Be infected with a virus that has transmission 14 and the voice change symptom.

### LINKED ISSUE #43677: "Severe damage to brain language center"cerebral trauma broken and won't fix when a surgery is done.
Issue reported from  Round ID: 106628 (/tg/Station Bagil [ENGLISH] [US-WEST] [100% FREE LAG])

Reporting client version: 512.1466
After taking a ton of meth I died and was cloned with the defects still on me .Then I was put into surgery to fix it and it did not fix, though the others did. The trauma makes anyone in chat appear to speak gibberish so it you can't understand it.

### LINKED ISSUE #47731: possible brain trauma+mutation toxin bug
https://cdn.discordapp.com/attachments/326826573951008769/643737815494033418/brain_damage_bug.PNG

I cloned the RD, and they had a lot of brain damage and multiple traumas, one of which kept them from speaking properly (I think it's the one that gives you the special brain damaged language and forces you to use (only) that language?). I later fed them podperson mutation toxin, which healed all of their organ damage (including their brain damage and their brain traumas, as indicated by the screenshot above). They were still just as unable to speak properly/Common as they were just after they were cloned. I forgot to ask them after the round if they were able to understand my speech (in Common).

Maybe the podperson mutation toxin destroyed the RD's brain traumas in a weird way, resulting in them not undoing their changes to his known languages?

### PR DESCRIPTION
**Pull request by kriskog**
## About The Pull Request
I just wanted to fix a couple bugs, but here we are.
Rewritten language holders to use a more sane structure, most of the management stuff is done automagically, and we got saner helpers in atom. A few new features snuck its way in along the way as well, like species language datums.

Key features:
 - The mob/mind duality of the old language holders has been thrown out in favor of sane priority. If a mind exists, its holder will be used and updated according to the mob/atom holder. If not, the mob/atom holder is used. This prioritization is handled in the atom procs, so in most cases there is no need to manually manage this.
 - Missing language holders are created if they are needed or about to be used, ie with say. No explicit creation of language holders before that with the exception of humans due to tying updates to set species.
 - Missing selected language is set when attempting to speak, based on language priority. Makes stuff like the voice change symptom actually work as intended.
 - Human species language holder support. Instead of adding languages during post_equip, we can now add species language holders to the species datum. Makes transforms give languages properly, and opens up for more easily adding race languages in future.
 - Helpers in atom/movable that handle everything, a lot less manual grabbing of language holders. 
 - Langauges (and blocks) have sources, which makes us much less prone to lose any we should still have.
 - Blocking languages replaces removing languages and juggling old and new language holders to revert. 
 - Comments.

To do:
- [x] Add proper autodoc.
- [x] Add spoken/understood option and better feedback to language menu.
- [x] ~Check holograms and imaginary friend behaviour. Doesn't seem to use language?~ Gone
- [x] Fix monkeyize

## Why It's Good For The Game
Fixes #25663
Fixes #27087
Fixes #29568
Fixes #37690
Fixes #42871
Fixes #43677
Fixes #47731

## Changelog
:cl: Skoglol
refactor: Refactored language holders. Fixes a bunch of language related bugs and inconsistencies.
fix: Transforms now apply new languages properly. Monkeys can no longer speak common.
add: Species changes will now give you the relevant languages.
/:cl:

### CODE REVIEW

File: code/datums/mind.dm [hunk 1]
```diff
@@ -128,6 +122,9 @@
 	if(active || force_key_move)
 		new_character.key = key		//now transfer the key to link the client to our new body
 
+	get_language_holder()			// Makes a language holder if one doesnt exist
+	language_holder.update_mob_languages(new_character)
```
Comment (AnturK) [inline code/datums/mind.dm]: You should use the returned value here for consistency.

File: code/game/objects/items/holy_weapons.dm [hunk 2]
```diff
@@ -452,7 +452,6 @@
 		S.ckey = C.ckey
 		S.fully_replace_character_name(null, "The spirit of [name]")
 		S.status_flags |= GODMODE
-		S.language_holder = user.language_holder.copy(S)
```
Comment (AnturK) [inline code/game/objects/items/holy_weapons.dm]: No replacement here ?

File: code/modules/language/language_holder.dm [hunk 3]
```diff
@@ -1,141 +1,226 @@
...
+			update_mob_languages(M.current)
+	get_selected_language()
 
 /datum/language_holder/Destroy()
-	owner = null
-	QDEL_NULL(language_menu)
-	languages.Cut()
-	shadow_languages.Cut()
+	qdel(language_menu)
+	qdel(temporary_languages)
```
Comment (AnturK) [inline code/modules/language/language_holder.dm]: Aren't these just lists of types ?

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by RaveRadbury** [comment]
> Makes transforms give languages properly
So if the station gets mass liz'd by a wiz event they'd also get draconic?

**Comment by kriskog** [comment]
Yes, so long as they are lizards theh would have access to draconic.

I need to redo this a bit but been a bit short on time. All languages need sources associated with them, so we dont keep losing languages we arent supposed to. Bear with me.

**Comment by RaveRadbury** [comment]
#26780 The original PR for draconic intended that it not be granted through transforms.

**Comment by kriskog** [comment]
> 
> 
> #26780 The original PR for draconic intended that it not be granted through transforms.

I refuse to snowflake in a system for non transform-granted languages. Mind you the old langauge holders did not have a species language system in place, so this may as well have been motivated by the path of least resistance as anything else.
It's an instinctual language now, just like any species granted language. Deal with it :slightly_smiling_face: 


(Updates still coming, just taking a while. No more reviews just yet)

File: code/game/objects/items/holy_weapons.dm (hunk 2 above)
Comment (kriskog) [reply code/game/objects/items/holy_weapons.dm]: This was in the known issues section, back in now. :)

**Comment by kriskog** [comment]
I still have a few small things on my to do, but this is review ready. No more big changes planned.

File: code/datums/dna.dm [hunk 4]
```diff
@@ -305,6 +305,11 @@
 		var/datum/species/old_species = dna.species
 		dna.species = new_race
 		dna.species.on_species_gain(src, old_species, pref_load)
+		qdel(language_holder)
+		var/species_holder = initial(mrace.species_language_holder)
+		language_holder = new species_holder(src)
+		if(mind)
```
Comment (AnturK) [inline code/datums/dna.dm]: Why the mind check here ?

File: code/modules/language/language_holder.dm [hunk 5]
```diff
@@ -1,141 +1,275 @@
...
-	var/only_speaks_language = null
-	var/selected_default_language = null
+	/// Understood languages.
+	var/list/understood_languages = list(/datum/language/common = list(LANGUAGE_MIND))
+	/// A list of languages that can be spoken. Tongue organ may also set limits beyond this list.
+	var/list/spoken_languages = list(/datum/language/common = list(LANGUAGE_ATOM))
+	/// A list of blocked languages. Used to prevent understanding and speaking certain languages, ie for certain mobs, mutations etc.
+	var/list/blocked_languages = list()
+	/// If true, overrides spoken_languages and tongue limitations.
+	var/omnitongue
```
Comment (AnturK) [inline code/modules/language/language_holder.dm]: default to FALSE if this is boolean

File: code/modules/language/language_holder.dm [hunk 6]
```diff
@@ -1,141 +1,275 @@
...
+		if(M.current)
+			update_atom_languages(M.current)
+	get_selected_language()
 
 /datum/language_holder/Destroy()
-	owner = null
-	QDEL_NULL(language_menu)
-	languages.Cut()
-	shadow_languages.Cut()
+	qdel(language_menu)
```
Comment (AnturK) [inline code/modules/language/language_holder.dm]: QDEL_NULL here

File: code/modules/language/language_holder.dm [hunk 7]
```diff
@@ -1,141 +1,275 @@
...
-/datum/language_holder/proc/grant_language(datum/language/dt, shadow = FALSE)
-	if(shadow)
-		shadow_languages[dt] = TRUE
-	else
-		languages[dt] = TRUE
+/// Grants the supplied language.
+/datum/language_holder/proc/grant_language(language, understood = TRUE, spoken = TRUE, source = LANGUAGE_MIND)
+	if(understood)
+		if(!understood_languages[language])
+			understood_languages += language
```
Comment (AnturK) [inline code/modules/language/language_holder.dm]: No need for this line. Same for spoken.

File: code/modules/language/language_holder.dm [hunk 8]
```diff
@@ -1,141 +1,275 @@
...
+				spoken_languages -= language
+		. = TRUE
 
-/datum/language_holder/proc/remove_all_languages()
-	languages.Cut()
+/// Removes every language and optionally sets omnitongue false. If a non default source is supplied, only removes that source.
+/datum/language_holder/proc/remove_all_languages(source = LANGUAGE_ALL, omni = FALSE)
+	for(var/language in GLOB.all_languages)
+		remove_language(language, TRUE, TRUE, source)
+	if(omni)
```
Comment (AnturK) [inline code/modules/language/language_holder.dm]: I'd rename parameter to remove_omni here so it's less confusing what result is

File: code/modules/language/language_holder.dm [hunk 9]
```diff
@@ -1,141 +1,275 @@
...
-			if(is_type_in_typecache(dt, L.shadow_languages))
-				return LANGUAGE_SHADOWED
-	return FALSE
+/// Adds a single language or list of languages to the blocked language list.
+/datum/language_holder/proc/add_blocked_language(languages, source = LANGUAGE_MIND)
+	if(!islist(languages))
+		languages = list(languages)
+	for(var/language in languages)
+		if(!blocked_languages[language])
+			blocked_languages += language
```
Comment (AnturK) [inline code/modules/language/language_holder.dm]: Same as before.

File: code/modules/language/language_holder.dm [hunk 10]
```diff
@@ -1,141 +1,275 @@
...
+/datum/language_holder/proc/can_speak_language(language, tongue = TRUE)
+	if((omnitongue || tongue) && has_language(language, TRUE))
+		return TRUE
+	return FALSE
 
-	if(replace)
-		src.remove_all_languages()
+/// Returns selected language if it can be spoken, or decides, sets and returns a new selected language if possible.
+/datum/language_holder/proc/get_selected_language()
+	if(selected_language && can_speak_language(selected_language))
```
Comment (AnturK) [inline code/modules/language/language_holder.dm]: Shouldn't this be taking tongue restrictions into account ?

**Review by AnturK — CHANGES_REQUESTED** [review]
When you read the word language enough times the meaning disappears.

**Comment by kriskog** [comment]
> When you read the word language enough times the meaning disappears.

Yeah no joke. Last time I did the math, language_holder.dm had a ratio of 1.35 language per line. Put me out of my misery.

File: code/datums/dna.dm (hunk 4 above)
Comment (kriskog) [reply code/datums/dna.dm]: Left over from old, mindless updates are handled in the proc now.

File: code/modules/language/language_holder.dm (hunk 10 above)
Comment (kriskog) [reply code/modules/language/language_holder.dm]: Yes, forgot that later in the proc too. Good catch

Comment (kriskog) [reply code/modules/language/language_holder.dm]: tongue check moved to can_speak_language

Comment (kriskog) [reply code/modules/language/language_holder.dm]: tongue check moved to can_speak_language

**Comment by kriskog** [comment]
Found another bug, for some reason the old language holder comes along on monkeyize now. Ill need to fix this as well.

**Comment by kriskog** [comment]
The language menu grant/remove changes were a bit bare bones, but it works and these things are rarely used anyway. I'll think up something better at a later point.

This should be good to go now.

**Pull request merged**